pmd-core/src/main/java/net/sourceforge/pmd/processor/PmdRunnable.java

```diff
@@ -81,7 +81,6 @@ public Report call() {
         }
 
         try (InputStream stream = new BufferedInputStream(dataSource.getInputStream())) {
-            tc.ruleContext.setLanguageVersion(null);
```

Comment:
I don't think actually works. The language version is reset to null here so that the SourceCodeProcessor can determine the language later in the call to processSourceCode. If you don't reset it, then even if you haven't specified a language on the command line, when processing the next file the LanguageVersionDiscoverer is not called again, since it sees the version is still there on the RuleContext. But that is the wrong version, since it's the one of the previous file.

See here:

https://github.com/pmd/pmd/blob/bc5da416aa355ee4177b474ce837ce2d6c069108/pmd-core/src/main/java/net/sourceforge/pmd/SourceCodeProcessor.java#L204-L210

I think you should keep this statement in, and instead edit the above function of SourceCodeProcessor to use Configuration#getInputLanguageVersion if it is provided

Reply:
I didn't like the idea of letting the PMDConfiguration class be visible all the way down in SourceCodeProcessor. So, I stored the initial language version before going into the loop over source files. Then it can be reset each time. If no initial language is supplied, then this will have the same effect as setting it directly to null. If an initial language was specified, that gets set back in on each iteration. 

Does that work for you?

Reply:
But it's already visible in SourceCodeProcessor, there's a field for it. SourceCodeProcessor will anyway be removed in pmd 7 so using this field is fine.

I still would prefer modifying the method of SourceCodeProcessor because it's not clear what you're resetting it to in PmdRunnable. I know it is either null, or the getInputLanguageVersion. But it's not written in the code and I'm going to forget it. Ultimately your PR changes the way we discover the language, so I think it's fitting to encapsulate that in the determineLanguage method

Reply:
Doh! I looked for it in SourceCodeProcessor but apparently looked right past it. Sorry about that. Yes, your suggestion makes more sense. I've changed my PR over to do it that way and squashed the commits together.

There's actually one more thing I want to check out with this before merging. So, with apologies for it being so noisy, please hang on before reviewing again.

Reply:
Hi @oowekyala, I've checked the thing I wanted to check: I wanted to do a proper test with the built version of PMD to make sure that it really does provide a way to address #2133. It does, indeed work for that. 

So, now it seems like I've arrived small and simple change in this PR. Please do review when you can.